### ISSUE TITLE
AssertionError: Torch not compiled with CUDA enabled

### ISSUE THREAD

**Issue by aktaseren** [pytorch/pytorch#50401]
## 🐛 Bug

```
~/Desktop/ULfAD/ceadarscript/DAGMMphase/model.py in compute_energy(self, z, phi, mu, cov, size_average)
    146         cov_inverse = torch.cat(cov_inverse, dim=0)
    147         # K
--> 148         det_cov = torch.cat(det_cov).cuda()
    149         #det_cov = to_var(torch.from_numpy(np.float32(np.array(det_cov))))
    150 

~/opt/anaconda3/lib/python3.7/site-packages/torch/cuda/__init__.py in _lazy_init()
    164                 "Cannot re-initialize CUDA in forked subprocess. " + msg)
    165         if not hasattr(torch._C, '_cuda_getDeviceCount'):
--> 166             raise AssertionError("Torch not compiled with CUDA enabled")
    167         if _cudart is None:
    168             raise AssertionError(
```

AssertionError: Torch not compiled with CUDA enabled

I am having the issue above while I run a script run by pytorch. I am using MacOS BigSur the latest version. My PyTorch version is 1.7.0. What do you suggest me to solve it?

cc @ngimel

**Comment by skyline75489** [comment]
The error message is clear. You’re not using a PyTorch with CUDA support.

**Comment by aktaseren** [comment]
@skyline75489 Sorry I forgot to mention. I use MacOS BigSur the latest version. Therefore, there is no Cuda applicability in my machine as far as I know. Is there any way to solve this?

**Comment by skyline75489** [comment]
I don’t know if you own the model code. Seems it deliberately uses cuda() which leads to the error. I don’t see another way to solve this other than changing the model code yourself

**Comment by aktaseren** [comment]
@skyline75489 the part of the code related to cuda is below. It is for the calculation of energy. Therefore, I guess it would be needed to use Cuda on det_cov high computation. However, if there is any way to use PyTorch without Cuda on MacOS BigSur,  that would be great to solve this step.
```

    def compute_energy(self, z, phi=None, mu=None, cov=None, size_average=True):
        if phi is None:
            phi = to_var(self.phi)
        if mu is None:
            mu = to_var(self.mu)
        if cov is None:
            cov = to_var(self.cov)

        k, D, _ = cov.size()

        z_mu = (z.unsqueeze(1)- mu.unsqueeze(0))

        cov_inverse = []
        det_cov = []
        cov_diag = 0
        eps = 1e-12
        for i in range(k):
            # K x D x D
            cov_k = cov[i] + to_var(torch.eye(D)*eps)
            cov_inverse.append(torch.inverse(cov_k).unsqueeze(0))

            #det_cov.append(np.linalg.det(cov_k.data.cpu().numpy()* (2*np.pi)))
            det_cov.append((Cholesky.apply(cov_k.cpu() * (2*np.pi)).diag().prod()).unsqueeze(0))
            cov_diag = cov_diag + torch.sum(1 / cov_k.diag())

        # K x D x D
        cov_inverse = torch.cat(cov_inverse, dim=0)
        # K
        det_cov = torch.cat(det_cov).cuda()
```

**Comment by skyline75489** [comment]
AFAIK, pytorch works well on macOS as long as you do not use cuda(). It will just fallback to CPU (or Metal? I don’t know if it’s available in 1.7)

获取 Outlook for iOS<https://aka.ms/o0ukef>
________________________________
发件人: Eren Aktas <notifications@github.com>
发送时间: Tuesday, January 12, 2021 8:27:50 PM
收件人: pytorch/pytorch <pytorch@noreply.github.com>
抄送: Chester Liu <skyline75489@outlook.com>; Mention <mention@noreply.github.com>
主题: Re: [pytorch/pytorch] AssertionError: Torch not compiled with CUDA enabled (#50401)


@skyline75489<https://github.com/skyline75489> the part of the code related to cuda is below. It is for the calculation of energy. Therefore, I guess it would be needed to use Cuda on det_cov high computation. However, if there is any way to use PyTorch without Cuda on MacOS BigSur, that would be great to solve this step.


    def compute_energy(self, z, phi=None, mu=None, cov=None, size_average=True):
        if phi is None:
            phi = to_var(self.phi)
        if mu is None:
            mu = to_var(self.mu)
        if cov is None:
            cov = to_var(self.cov)

        k, D, _ = cov.size()

        z_mu = (z.unsqueeze(1)- mu.unsqueeze(0))

        cov_inverse = []
        det_cov = []
        cov_diag = 0
        eps = 1e-12
        for i in range(k):
            # K x D x D
            cov_k = cov[i] + to_var(torch.eye(D)*eps)
            cov_inverse.append(torch.inverse(cov_k).unsqueeze(0))

            #det_cov.append(np.linalg.det(cov_k.data.cpu().numpy()* (2*np.pi)))
            det_cov.append((Cholesky.apply(cov_k.cpu() * (2*np.pi)).diag().prod()).unsqueeze(0))
            cov_diag = cov_diag + torch.sum(1 / cov_k.diag())

        # K x D x D
        cov_inverse = torch.cat(cov_inverse, dim=0)
        # K
        det_cov = torch.cat(det_cov).cuda()


―

**Comment by aktaseren** [comment]
@skyline75489 I updated the line triggering Cuda to not use Cuda since it is not working on MacOS as follows:

det_cov = to_var(torch.from_numpy(np.float32(np.array(det_cov))))

The script is working with PyTorch now.
Thanks.
